content/post/date_features.Rmd

```diff
@@ -0,0 +1,184 @@
+---
+title: Feature Engineering of Date-Time Variables
+author: Lennart Schneider
+date: '2020-02-20'
+slug: date-features
+categories: []
+tags: ['feature-engineering', 'mlr3pipelines', 'random-forest', 'regression']
+packages: ['mlr3', 'mlr3learners', 'mlr3pipelines', 'ranger']
+---
+
+```{r, echo=FALSE}
+library(mlr3book)
+```
+
+In this tutorial, we demonstrate how `r mlr_pkg("mlr3pipelines")` can be used to easily engineer
+features based on date-time variables. Relying on the [Bike Sharing
+Dataset](https://archive.ics.uci.edu/ml/datasets/bike+sharing+dataset) and the
+`r ref("mlr_learners_regr.ranger", "ranger learner")` we compare the RMSE of a random forest using
+the original features (baseline), to the RMSE of a random forest using newly engineered features on
+top of the original ones.
+
+## Motivation
+A single Date-Time variable (i.e., a `POSIXct` column) contains plenty of information ranging from
+year, month, day, hour, minute and second to other features such as week of the year, or day of the
+week. Moreover, most of these features are of cyclical nature, i.e., the eleventh and twelfth hour
+of a day are one hour apart, but so are the 23rd hour and midnight of the other day (see also this
+[blog post](http://blog.davidkaleko.com/feature-engineering-cyclical-features.html) and
+[fastai](https://docs.fast.ai/tabular.transform.html#Treating-date-columns) for more information).
+
+Not respecting this cyclical nature results in treating hours on a linear continuum. One way to
+handle a cyclical feature, $\mathbf{x}$, is to compute the sines and cosines transformation of
+$\frac{2 \pi \mathbf{x}}{max_{\mathbf{x}}}$. This results in a two-dimensional representation of the
+feature:
+```{r, echo=FALSE, fig.height=5.5, fig.width=10, fig.align = "center"}
+hours = 0:23
+hours_scaled = (2 * pi * hours) / 24
+hours_sin = sin(hours_scaled)
+hours_cos = cos(hours_scaled)
+n = 24
+cols = rainbow(n, s = 1, v = 1, start = 0, end = max(1, n - 1) / n, alpha = 1)
+par(mfrow = c(1, 2))
+plot(hours, type = "n", xlab = "Index", ylab = "Hours")
+text(hours, col = cols, labels = as.character(hours))
+plot(hours_sin, hours_cos, type = "n", xlab = "Sinus Transformation", ylab = "Cosines Transformation")
+text(hours_sin, hours_cos, col = cols, labels = as.character(hours))
+```
+
+`r mlr_pkg("mlr3pipelines")` provides the `PipeOpDateFeatures` pipeline which can be used to
+automatically engineer features based on `POSIXct` columns, including handling of cyclical features.
+
```

Comment:
Maybe also mention, that most learners can not naturally handle dates / posixct data, and therefore conversion is often naturally required before beeing able to train anything.